### PR TITLE
[DO NOT MERGE][NCCL][CUDA][CUDA Graphs] Set watchdog runtime capture mode to thread local to handle cleaning straggling work

### PR DESCRIPTION
**Pull request by eqy**
A more minimal alternative to #104487 and #104555 that should not unwittingly suppress errors during graph captures.

I've tested this locally and it appears to fix the repro used to test #104487 and #104555, while checking that the watchdog is attempting to call `cudaEventQuery` (on an event recorded before the capture) in the middle of a capture on another thread.

X-posting summary from #104555 here:
1. The issue is a crash caused by the watchdog thread calling `cudaEventQuery` on work that was enqueued before a graph capture.
2. The initial proposal was to change the capture mode of the thread performing the capture to the "thread local" mode, so that other threads querying events (the watchdog) don't crash as the events in question were created before the actual capture and therefore shouldn't interfere with the capture.
3. Many such as @eellison and @albanD were concerned with how we changed the capture mode of the capturing thread in this PR (by globally changing the constructor), as we could potentially miss errors/failures caused by always using this more lenient mode.
4. @albanD pointed out that it is really just the watchdog thread that we are concerned with issuing a disallowed `cudaEventQuery` call, so another potential solution is to just have the watchdog thread operate in a more lenient mode. This seems to be fairly minimal change that shouldn't have unexpected effects on the correctness or debuggability of code involving graph captures. To clarify, while the watchdog thread itself is not capturing anything, since the watchdog thread never calls `cudaThreadExchangeStreamCaptureMode`, it is operating with the "global" mode when another thread is in the middle of a capture, and hence crash the capture because `cudaEventQuery` is disallowed in this scenario.
4a. Our current understanding is the following (assuming the `cudaEventQuery` is for an event outside of a capture).

capturing thread mode | watchdog thread mode | watchdog event query allowed |
--------------------------|---------------------------|-----------------------------------|
global (default)             | global (default)              | no                                                |
global (default)             | thread local                    | yes  (this PR)                               |
thread local                   | global (default)              | yes  (#104555)                             |
thread local                   |thread local                     | yes                                              |

CC @kwen2501 @eellison @albanD @ptrblck @Aidyn-A 

cc @ptrblck @mcarilli @ezyang

### CODE REVIEW

**Review by kwen2501 — APPROVED** [review]
Glad to know that this solution also works! Thanks for the detailed analysis!

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -825,6 +825,8 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 void ProcessGroupNCCL::ncclCommWatchdog() {
   try {
     LOG(INFO) << "[Rank " << rank_ << "] NCCL watchdog thread started!";
+    auto mode = cudaStreamCaptureModeThreadLocal;
```
Comment (eellison) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:828]: do we need to set back the thread to its previous mode after this is done ?

**Review by eellison — COMMENTED** [review]
cool! great to find a better solution here. one question

**Review by albanD — APPROVED** [review]
Looks great!
I guess you don't need to change it back since this function is only executed in the worker thread that does only this?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (eqy) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:828]: I wouldn't think so unless the watchdog thread is further spawning additional threads or doing something other than simply polling enqueued work.

**Comment by eqy** [comment]
A distributed test seems to be timing out: https://pipelines.actions.githubusercontent.com/serviceHosts/7d146c05-69c3-4c20-a0e7-818111670117/_apis/pipelines/1/runs/3158352/signedlogcontent/876?urlExpires=2023-07-15T03%3A55%3A24.7602054Z&urlSigningMethod=HMACV1&urlSignature=EgXarp8aALtIP9Gt18kd3JOkndAiRbEo5y%2B37%2Fi4kBA%3D
adding a `C10_CUDA_CHECK` and restoring the original mode to see if that makes any difference...

**Comment by eqy** [comment]
Testing a theory now that somehow the CUDA runtime API call in the watchdog is causing the M60 runner to time out on this distributed test.

**Comment by ruanslv** [comment]
Any theories for the failure?

**Comment by eqy** [comment]
#105625 hopefully improves the RTT on test timing-out/feedback:

**Comment by bottler** [comment]
@eqy Do you want to do a manual rebase of this? The test failure is to do with SyncBatchNorm and seems interesting.

**Comment by eqy** [comment]
Rebased, sorry for the noise around this

**Comment by mikelambert** [comment]
Thank you, it seems this PR fixes the issue seen we've seen in https://github.com/pytorch/pytorch/pull/105182 ! Can you expand on the meaning of the `[DO NOT MERGE]` on this PR, and whether we should treat it as a warning against pulling it into our 2.1.0 build to fix the issues we're encountering? Thanks!

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -427,21 +427,30 @@ bool ProcessGroupNCCL::WorkNCCL::startedGPUExecutionInternal() const {
...
       // Checking the work's corresponding CUDA events' status
       if (!(*ncclEndEvents_)[i].query()) {
+        // pop
+        cudaThreadExchangeStreamCaptureMode(&mode);
         return false;
       }
     }
   } catch (const std::exception& e) {
+    // pop
+    cudaThreadExchangeStreamCaptureMode(&mode);
```
Comment (mikelambert) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:444]: If this catch branch runs and doesn't re-throw, don't we end up double-popping (aka leaving it as 'local') when `finishedGPUExecutionInternal` finishes?

**Pull request closed without merging**